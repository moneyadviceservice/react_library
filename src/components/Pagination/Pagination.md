Pagination Component

```jsx
import React, { useState } from 'react';
import { Pagination } from '@moneypensionservice/directories';

const items = ['item1', 'item2', 'item3', 'item4', 'item5'];

const [currentPage, setPage] = useState(2);
const itemsPerPage = 2;
const nPages = Math.ceil(items.length / itemsPerPage);

<Pagination 
  currentLng="en"
  currentPage={currentPage} 
  totalPages={nPages} 
  nextClick={() => setPage(currentPage + 1)}
  prevClick={() => setPage(currentPage - 1)} />
```
