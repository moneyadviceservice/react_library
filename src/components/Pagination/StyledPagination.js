import styled from 'styled-components'
import Button from '../Button'
import { Col } from '../Grid'
import { Paragraph } from '../Paragraph'

const StyledPagination = styled(Col)`
  ${props => !props.margin && `margin: 0 auto;`}
`
const PaginationButton = styled(Button)`
  ${props => props.noShow && `display: none;}`}
`
const PaginationCounter = styled(Paragraph)`
  margin: 0;
  width: auto;
  padding: 0 20px;
`

export { StyledPagination, PaginationButton, PaginationCounter }
