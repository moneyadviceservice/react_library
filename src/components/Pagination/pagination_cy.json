{
  "previous": "Flaenorol",
  "next": "Nesaf",
  "page": "Tudalen",
  "of": "o"
}
