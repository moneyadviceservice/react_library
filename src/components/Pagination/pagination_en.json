{
  "previous": "Prev",
  "next": "Next",
  "page": "Page",
  "of": "of"
}
