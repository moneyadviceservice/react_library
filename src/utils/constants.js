export const breakpoints = {
  xs: 1,
  sm: 48, // 768px
  md: 64, // 1024px
  lg: 90, // 1440px
  xl: 120, // 1920px
}

export const dimensions = ['xs', 'sm', 'md', 'lg', 'xl']

export const mediaQuery = 'only screen'

export const ALIGN_SELF_MAP = {
  center: 'center',
  end: 'flex-end',
  start: 'flex-start',
  stretch: 'stretch',
}

export const SPACING_SIZES = [
  'xxsmall',
  'xsmall',
  'small',
  'medium',
  'large',
  'xlarge',
]

