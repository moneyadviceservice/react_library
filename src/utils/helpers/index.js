export * from './background'
export * from './responsive'
export * from './flexbox'
export * from './styles'
export * from './mediaQueries'
export * from './sizing'
export * from './refs'
