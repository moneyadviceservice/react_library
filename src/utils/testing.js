import { render, act } from '@testing-library/react'
import GlobalTheme from '../components/ThemeProvider'

const customRender = (ui, options) =>
  render(ui, { wrapper: GlobalTheme, ...options })

const waitRender = async (ui, options) => {
  const { container, ...rest } = customRender(ui, options)

  await act(async () => {
    await Promise.resolve(container)
  })

  return { container, ...rest }
}

export * from '@testing-library/react'

export { customRender as render, waitRender }
