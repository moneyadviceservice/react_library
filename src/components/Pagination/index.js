import React from 'react'
import PropTypes from 'prop-types'
import { genericPropTypes, genericPropsDefaults } from '../../utils/prop-types'
import {
  StyledPagination,
  PaginationButton,
  PaginationCounter,
} from './StyledPagination'
// translations
import LocaleEn from './pagination_en.json'
import LocaleCy from './pagination_cy.json'

const Pagination = ({
  a11yTitle,
  currentLng,
  currentPage,
  nextClick,
  nextUrl,
  prevClick,
  prevUrl,
  totalPages,
  ...rest
}) => {
  const i18n = currentLng === 'cy' ? LocaleCy : LocaleEn

  return (
    <StyledPagination
      aria-label={a11yTitle}
      forwardedAs="nav"
      direction="row"
      align="center"
      justify="center"
      {...rest}
    >
      <PaginationButton
        text={`\u003C ${i18n.previous}`}
        weight={400}
        href={prevUrl}
        onClick={prevClick}
        noShow={currentPage === 1}
      />
      <PaginationCounter>
        {`${i18n.page} ${currentPage} ${i18n.of} ${totalPages}`}
      </PaginationCounter>
      <PaginationButton
        text={`${i18n.next} \u003E`}
        weight={400}
        href={nextUrl}
        onClick={nextClick}
        noShow={currentPage === totalPages}
      />
    </StyledPagination>
  )
}

// Documentation
Pagination.propTypes = {
  /** Current Language Value */
  currentLng: PropTypes.oneOf(['en', 'cy']),
  /** Sets the currentPage being displayed. */
  currentPage: PropTypes.number,
  /** Function to be triggered by the next button. */
  nextClick: PropTypes.func,
  /** Sets a url on the next button. Button will render as an anchor. */
  nextUrl: PropTypes.string,
  /** Function to be triggered by the prev button. */
  prevClick: PropTypes.func,
  /** Sets a url on the prev button. Button will render as an anchor. */
  prevUrl: PropTypes.string,
  /** Sets the total pages to display. */
  totalPages: PropTypes.number,
  ...genericPropTypes,
}

Pagination.defaultProps = {
  currentLng: 'en',
  ...genericPropsDefaults(),
}

/** @component */
export { Pagination }
