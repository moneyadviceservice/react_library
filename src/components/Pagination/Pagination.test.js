import React from 'react'
import { render, fireEvent } from '../../utils/testing'
import { Pagination } from '../Pagination'

describe('<Pagination />', () => {
  describe('Snapshots', () => {
    it('should match snapshot', () => {
      const { container } = render(
        <Pagination currentPage={2} totalPages={3} />
      )
      expect(container).toMatchSnapshot()
    })
    it('should match snapshot in Welsh', () => {
      const { container } = render(
        <Pagination currentPage={2} totalPages={3} currentLng="cy" />
      )
      expect(container).toMatchSnapshot()
    })
    it('should match snapshot in first page', () => {
      const { container } = render(
        <Pagination currentPage={1} totalPages={2} />
      )
      expect(container).toMatchSnapshot()
    })
    it('should match snapshot in last page', () => {
      const { container } = render(
        <Pagination currentPage={3} totalPages={3} />
      )
      expect(container).toMatchSnapshot()
    })
  })

  describe('nextClick prop', () => {
    it('should call nextClick function when right button is clicked', () => {
      const nextClickMock = jest.fn()
      const { getByTestId } = render(
        <Pagination currentPage={2} totalPages={3} nextClick={nextClickMock} />
      )

      fireEvent.click(getByTestId('nextBtn'))

      expect(nextClickMock).toHaveBeenCalledTimes(1)
    })
  })

  describe('prevClick prop', () => {
    it('should call prevClick function when left button is clicked', () => {
      const prevClickMock = jest.fn()
      const { getByTestId } = render(
        <Pagination currentPage={2} totalPages={3} prevClick={prevClickMock} />
      )

      fireEvent.click(getByTestId('prevBtn'))

      expect(prevClickMock).toHaveBeenCalledTimes(1)
    })
  })

  describe('nextUrl prop', () => {
    it('should apply href property to right button', () => {
      const { container, getByTestId } = render(
        <Pagination
          currentPage={2}
          totalPages={3}
          nextUrl="https://www.test.com/"
        />
      )
      // snapshot with url
      expect(container).toMatchSnapshot()
      // href
      expect(getByTestId('nextBtn').href).toBe('https://www.test.com/')
    })
  })

  describe('prevUrl prop', () => {
    it('should apply href property to left button', () => {
      const { container, getByTestId } = render(
        <Pagination
          currentPage={2}
          totalPages={3}
          prevUrl="https://www.test.com/"
        />
      )
      // snapshot with url
      expect(container).toMatchSnapshot()
      // href
      expect(getByTestId('prevBtn').href).toBe('https://www.test.com/')
    })
  })
})
