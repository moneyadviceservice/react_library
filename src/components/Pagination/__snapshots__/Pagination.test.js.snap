// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> Snapshots should match snapshot 1`] = `
<div>
  <nav
    class="styledCol__ColWrapper-sc-1mivnjm-0 gaLJSE StyledPagination-hmwnoe-0 cmeGul"
  >
    <button
      class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 jEdQwJ StyledPagination__PaginationButton-hmwnoe-1"
      data-testid="prevBtn"
      type="button"
    >
      &lt; Prev
    </button>
    <p
      class="StyledParagraph-sc-1toaej-0 gpltyO StyledPagination__PaginationCounter-hmwnoe-2 dMIRzi"
    >
      Page 2 of 3
    </p>
    <button
      class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 jEdQwJ StyledPagination__PaginationButton-hmwnoe-1"
      data-testid="nextBtn"
      type="button"
    >
      Next &gt;
    </button>
  </nav>
</div>
`;

exports[`<Pagination /> Snapshots should match snapshot in Welsh 1`] = `
<div>
  <nav
    class="styledCol__ColWrapper-sc-1mivnjm-0 gaLJSE StyledPagination-hmwnoe-0 cmeGul"
  >
    <button
      class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 jEdQwJ StyledPagination__PaginationButton-hmwnoe-1"
      data-testid="prevBtn"
      type="button"
    >
      &lt; Flaenorol
    </button>
    <p
      class="StyledParagraph-sc-1toaej-0 gpltyO StyledPagination__PaginationCounter-hmwnoe-2 dMIRzi"
    >
      Tudalen 2 o 3
    </p>
    <button
      class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 jEdQwJ StyledPagination__PaginationButton-hmwnoe-1"
      data-testid="nextBtn"
      type="button"
    >
      Nesaf &gt;
    </button>
  </nav>
</div>
`;

exports[`<Pagination /> Snapshots should match snapshot in first page 1`] = `
<div>
  <nav
    class="styledCol__ColWrapper-sc-1mivnjm-0 gaLJSE StyledPagination-hmwnoe-0 cmeGul"
  >
    <button
      class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 jEdQwJ StyledPagination__PaginationButton-hmwnoe-1 cOvXO"
      data-testid="prevBtn"
      type="button"
    >
      &lt; Prev
    </button>
    <p
      class="StyledParagraph-sc-1toaej-0 gpltyO StyledPagination__PaginationCounter-hmwnoe-2 dMIRzi"
    >
      Page 1 of 2
    </p>
    <button
      class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 jEdQwJ StyledPagination__PaginationButton-hmwnoe-1"
      data-testid="nextBtn"
      type="button"
    >
      Next &gt;
    </button>
  </nav>
</div>
`;

exports[`<Pagination /> Snapshots should match snapshot in last page 1`] = `
<div>
  <nav
    class="styledCol__ColWrapper-sc-1mivnjm-0 gaLJSE StyledPagination-hmwnoe-0 cmeGul"
  >
    <button
      class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 jEdQwJ StyledPagination__PaginationButton-hmwnoe-1"
      data-testid="prevBtn"
      type="button"
    >
      &lt; Prev
    </button>
    <p
      class="StyledParagraph-sc-1toaej-0 gpltyO StyledPagination__PaginationCounter-hmwnoe-2 dMIRzi"
    >
      Page 3 of 3
    </p>
    <button
      class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 jEdQwJ StyledPagination__PaginationButton-hmwnoe-1 cOvXO"
      data-testid="nextBtn"
      type="button"
    >
      Next &gt;
    </button>
  </nav>
</div>
`;

exports[`<Pagination /> nextUrl prop should apply href property to right button 1`] = `
<div>
  <nav
    class="styledCol__ColWrapper-sc-1mivnjm-0 gaLJSE StyledPagination-hmwnoe-0 cmeGul"
  >
    <button
      class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 jEdQwJ StyledPagination__PaginationButton-hmwnoe-1"
      data-testid="prevBtn"
      type="button"
    >
      &lt; Prev
    </button>
    <p
      class="StyledParagraph-sc-1toaej-0 gpltyO StyledPagination__PaginationCounter-hmwnoe-2 dMIRzi"
    >
      Page 2 of 3
    </p>
    <a
      class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 jEdQwJ StyledPagination__PaginationButton-hmwnoe-1"
      data-testid="nextBtn"
      href="https://www.test.com/"
    >
      Next &gt;
    </a>
  </nav>
</div>
`;

exports[`<Pagination /> prevUrl prop should apply href property to left button 1`] = `
<div>
  <nav
    class="styledCol__ColWrapper-sc-1mivnjm-0 gaLJSE StyledPagination-hmwnoe-0 cmeGul"
  >
    <a
      class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 jEdQwJ StyledPagination__PaginationButton-hmwnoe-1"
      data-testid="prevBtn"
      href="https://www.test.com/"
    >
      &lt; Prev
    </a>
    <p
      class="StyledParagraph-sc-1toaej-0 gpltyO StyledPagination__PaginationCounter-hmwnoe-2 dMIRzi"
    >
      Page 2 of 3
    </p>
    <button
      class="buttonStyles__ButtonWrapper-sc-1c1ve7f-0 jEdQwJ StyledPagination__PaginationButton-hmwnoe-1"
      data-testid="nextBtn"
      type="button"
    >
      Next &gt;
    </button>
  </nav>
</div>
`;
